src/shims/fs.rs

```diff
@@ -257,6 +264,196 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                        if pathname.as_os_str().is_empty() && (flags & at_empty_path != 0) {
+                            (*path).clone()
+                        } else {
+                            // It is not specified what should happen here.
+                            throw_unsup_format!(
+                                "the file descriptor `{}` and path `{:?}` combination is invalid.",
+                                dirfd,
+                                path
+                            );
+                        }
+                    }
+                } else {
+                    return this.handle_not_found();
+                }
+            }
+        };
+
+        // the `_mask_op` paramter specifies the file information that the caller requested.
+        // However `statx` is allowed to return information that was not requested or to not
+        // return information that was requested. This `mask` represents the information we can
+        // actually provide in any host platform.
+        let mut mask =
+            this.eval_libc("STATX_TYPE")?.to_u32()? | this.eval_libc("STATX_SIZE")?.to_u32()?;
+
+        // If the `AT_SYMLINK_NOFOLLOW` flag is set, we query the file's metadata without following
+        // symbolic links.
+        let metadata = if flags & this.eval_libc("AT_SYMLINK_NOFOLLOW")?.to_i32()? != 0 {
+            std::fs::symlink_metadata(path)
+        } else {
+            std::fs::metadata(path)
+        };
+
+        let metadata = match metadata {
+            Ok(metadata) => metadata,
+            Err(e) => {
+                this.set_last_error_from_io_error(e)?;
+                return Ok(-1);
+            }
+        };
+
+        let file_type = metadata.file_type();
+
+        // The `mode` field specifies the type of the file and the permissions over the file for
+        // the owner, its group and other users. Given that we can only provide the file type
+        // without using platform specific methods, we only set the bits corresponding to the file
+        // type.
+        let mode = if file_type.is_file() {
+            this.eval_libc("S_IFREG")?
+        } else if file_type.is_dir() {
+            this.eval_libc("S_IFDIR")?
+        } else {
+            this.eval_libc("S_IFLNK")?
+        }
+        .to_u32()? as u16;
+
+        let size = metadata.len();
+
+        let (access_sec, access_nsec) = extract_sec_and_nsec(
+            metadata.accessed(),
+            &mut mask,
+            this.eval_libc("STATX_ATIME")?.to_u32()?
+        )?;
+
+        let (created_sec, created_nsec) = extract_sec_and_nsec(
+            metadata.created(),
+            &mut mask,
+            this.eval_libc("STATX_BTIME")?.to_u32()?
+        )?;
+
+        let (modified_sec, modified_nsec) = extract_sec_and_nsec(
+            metadata.modified(),
+            &mut mask,
+            this.eval_libc("STATX_MTIME")?.to_u32()?
+        )?;
+
+        let __u32_layout = this.libc_ty_layout("__u32")?;
```

Comment:
Isn't this just `layout_of(tcx.types.u32)` and friends?

Reply:
They are but I wanted to keep the same names as `man`.

Reply:
The manpage certainly doesn't use Rust types like `u32` though... oh, actually it does. Well never mind then. ;)